Return concrete Hash from hash constructors (#21)
microsoft/go-crypto-openssl: docs/go-openssl-compat.md

# Compatibility of Go BoringCrypto and OpenSSL

## Goal

List implementation decisions taken to smooth out differences between Go BoringCrypto expectations and OpenSSL behavior.

## Background

> Go BoringCrypto branch delegates core cryptographic algorithms to BoringSSL. Go does not support pluggable crypto backends, so in order to support BoringSSL, the BoringCrypto branch contains non-trivial modifications to several `crypto` packages. These modifications are not generic abstractions over crypto algorithms as they specifically target BoringSSL compatibility.

Author: Agree!

In this situations were there is no direct mapping, we try to combine several OpenSSL functions so the security is not compromised, being the trade-off between speed and maintainability.

## Implementation decisions

### AES-GCM encryption in TLS mode

#### Background

AES-GCM with Additional Data (aka AEAD) is a symmetric block cipher whose text-book encryption inputs are:
- initialization vector (aka nonce or iv)
- plaintext
- additional authenticated data

And the outputs are:
- ciphertext
- authentication tag

Go abstracts this algorithm using the `Seal` method in the [cipher.AEAD](https://pkg.go.dev/crypto/cipher#AEAD) interface:

`Seal(dst, nonce, plaintext, additionalData []byte) []byte`

Where dst is just a backing memory buffer to reduce allocations and the returned byte slice is the ciphertext with the authentication tag appended.

> When AES-GCM is used to encrypt a TLS payload (aka TLS mode), FIPS 140-2 IG A.5 ["Key/IV Pair Uniqueness * Requirements from SP 800-38D"](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf), and the general security consensus, requires constructing the IV parameter deterministically by concatenating two fields:

Author: This paragraph is describing how to construct an IV concatenating the fixed and the explicit field as per 800-38D 8.3, I think having a link to that spec here is helpful.

I guess you are confused because I do not mention the 2^64-1 limit imposed by FIPS 140-2 IG in the background section, which I should.

I will add a new paragraph mentioning it and referencing the IG 😸

> When AES-GCM is used to encrypt a TLS payload (aka TLS mode), FIPS 140-2 IG A.5 ["Key/IV Pair Uniqueness * Requirements from SP 800-38D"](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf), and the general security consensus, requires constructing the IV parameter deterministically by concatenating two fields:

Author: Also, notice that 800-38D 8.3 only imposes the 2^32 counter limit if the IV length is not 92bits (12 bytes), or when using RBG construction:

> The following requirement applies to all implementations that use either 1) the deterministic 
construction with IVs whose length is not 96, or 2) the RBG-based construction, for IVs of any 
length. In other words, unless an implementation only uses 96-bit IVs that are generated by the 
deterministic construction:

We only accept 12 bytes IVs, so this stricter limit does not apply to us.

> When AES-GCM is used to encrypt a TLS payload (aka TLS mode), FIPS 140-2 IG A.5 ["Key/IV Pair Uniqueness * Requirements from SP 800-38D"](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf), and the general security consensus, requires constructing the IV parameter deterministically by concatenating two fields:

Reviewer: Perfect, that clears it all up, thanks!

> The [FIPS 140-2 Implementation Guidance)(https://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf) requires that the explicit IV counter does not exceed the 2^64-1 value.

Reviewer 2: ```suggestion
The [FIPS 140-2 Implementation Guidance](https://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf) requires that the explicit IV counter does not exceed the 2^64-1 value.
```

    - Construct the AES-GCM TLS cipher using the new method.
    - Don't generate an IV when the cipher implements the AEAD2 interface.
    - Use the AEAD2 interface if available when encrypting.
    - Test and validate the new path.

2. Use OpenSSL AES-GCM in non-TLS mode, which means Go constructs the IV and OpenSSL honors it, but then we miss some required security checks that we would have to implement ourselves, namely:
  - Enforce strictly monotonically increasing explicit nonces.
  - Enforce explicit nonce values to be less than 2^64 - 1.
  - Tets the new checks.

#### Resolution

We will implement option 2.

Reasoning:
- The additional security checks are easy to implement.
- The changes are contained in `go-crypto-openssl`.
- Option 1 would require non-trivial patching of the Go TLS stack, which we would need to keep up-to-date when `crypto/tls` changes.
- Option 1 has higher chances to introduce a behavior chance or a security bug.

Drawbacks:
- OpenSSL might implement additional IV checks in future versions when running in TLS mode, and we will not benefit from them. We will have to keep an eye out for changes in this area.

> - We will implement security checks that go outside the ideal scope of `go-crypto-openssl`, which is just to act at as a thin layer between Go and OpenSSL APIs without real security knowledge.

Reviewer: I think simplifying makes it clearer on first read that the `which` is about the scope, not the security checks we're implementing:

```suggestion
- Adding security checks goes outside the ideal scope of `go-crypto-openssl`, which is just to act at as a thin layer between Go and OpenSSL APIs without real security knowledge.
```
